Introduce baseline JIT compiler

qwe661234 opened the pull request:
The baseline JIT compiler design involves tracing an EBB when its usage frequency exceeds a predetermined threshold. We then utilize a code generator to convert the instruction sequence into C code. The C code is subsequently compiled using mir, and the resulting target machine code is stored in the code cache for future utilization.

The primary objective of introducing the baseline JIT compiler is to enhance the execution speed of RISC-V instructions with minimal modification. This implementation requires three additional components: a code generator, mir compiler, and code cache, while maintaining the original design of the interpreter. Furthermore, this baseline JIT compiler serves as the foundational target for future improvements.

In addition, we have developed a Python script that effectively traces code templates and automatically generates JIT code templates. This approach eliminates the need for manually writing duplicated codey.

* Core i7-8700

| Metric    |     master      |  baseline JIT   |Speedup |
|-----------|-----------------|-----------------|--------|
| CoreMark  |1352.843 (iter/s)|1932.496 (iter/s)| +42.8% |
| Dhrystone |1282 DMIPS       |2495.11 DMIPS    | +94.6% |

* eMAG

| Metric    |     master      |  baseline JIT   |Speedup |
|-----------|-----------------|----------------|--------|
| CoreMark  | 438.073 (iter/s)| 644.426 (iter/s)| +47.1% |
| Dhrystone | 354.60 DMIPS    | 867.00 DMIPS    |  +144% |

Known issue: this implementation does not correctly work on Apple M1 MacOS.

jserv committed:
Decouple RV32 instructions as implementation templates

This commit prepares the groundwork for the future development of
a baseline JIT compiler. By decoupling the RV32 instructions, we
create reusable implementation templates that can be leveraged for
code generation in the JIT compiler.

This decoupling enables straightforward and flexible code generation
strategies, without code duplication between interpreter and MIR
based JIT compiler.

jserv reviewed (changes requested):
You shall rework the commit "Decouple RV32 instructions as implementation templates" and then rebase. i.e., at least two commits.

jserv reviewed (changes requested):
Explicitly mention commit 54024f74d7f1b86bddae0b011bd0469e1ba378db while comparing with JIT compilation in git commit message.

jserv commented on src/emulate.c:
```diff
@@ -310,971 +171,10 @@
-        rv->X[fuse[i].rd] = rv->io.mem_read_w(rv, addr);
-    }
-})
+/* FIXME: Add JIT-based execution path */
```
Remove this comment.

jserv commented on src/rv32_template.c:
```diff
@@ -0,0 +1,946 @@
+ * exception flag if either input is a signaling NaN.
+ */
+RVOP(feqs, {
+    rv->X[ir->rd] = (rv->F[ir->rs1] == rv->F[ir->rs2]) ? 1 : 0;
```
@qwe661234, let's create another pull request for fixing this.

jserv commented on src/cache.h:
```diff
@@ -37,4 +38,8 @@ void *cache_put(struct cache *cache, uint32_t key, void *value);
+void cache_free(struct cache *cache, void (*callback)(void *));
+
+#if RV32_HAS(JIT)
+bool cache_hot(struct cache *cache, uint32_t key);
```
Add comments here.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,411 @@
+    void *func;
+} func_obj_t;
+
+#define SET_SIZE_BITS 10
```
Add some comments here, illustrating set data structure.

jserv replied:
?

qwe661234 replied:
I add comments of set structure below.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,411 @@
+static riscv_jit_t *jit = NULL;
+static jit_item_t *jit_ptr = NULL;
+
+/* TODO: fix error when running on ARM architecture */
```
You shall describe the problem statement in details.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,430 @@
+    char *code;
+    size_t code_size;
+    size_t curr;
+} jit_item_t;
```
I am afraid that `jit_item_t` is a bit misleading. Did you mean that it is a matter of code cache?

qwe661234 replied:
This is a data structure used by c2mir to retrieve the code string.

qwe661234 replied:
Maybe we can rename it to `jit_gencode_t`?

jserv replied:
How about `code_string_t`? By the way, you shall provide more relevant comments.

jserv commented on src/riscv_private.h:
```diff
@@ -12,7 +12,9 @@
-
+#if RV32_HAS(JIT)
+#include "cache.h"
+#endif
```
Append a new line right after `#endif`.

jserv commented on src/riscv_private.h:
```diff
@@ -119,8 +126,13 @@ struct riscv_internal {
+#if !RV32_HAS(JIT)
     block_map_t block_map; /**< basic block map */
+#else
+    struct cache *cache;
```
The naming of this object `cache` is misleading. Can you clarify?

jserv commented on src/riscv_private.h:
```diff
@@ -59,8 +61,12 @@ typedef struct block {
+#endif
 } block_t;
 
+#if !RV32_HAS(JIT)
```
Drop this preprocessor directive since it is a well-defined declaration. Keeping it doesn't hurt.

jserv replied:
Add comments to annotate the functionality of this struct.

jserv commented on src/riscv_private.h:
```diff
@@ -59,8 +62,13 @@ typedef struct block {
 } block_t;
 
+/* map the translated basic block to its program counter  to accelerate
+ * searching. */
```
Mention that this struct is used only in interpreter at the moment.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,435 @@
+
+static int get_string_func(void *data)
+{
+    code_string_t *jit_code_string = data;
```
Replace `jit_code_string` with `codestr` for short.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,435 @@
+typedef struct {
+    char *code;
+    size_t code_size;
+    size_t curr;
```
You shall explain the functionality of `curr`.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,435 @@
+    __builtin___clear_cache((char *) (addr), (char *) (addr) + (size));
+
+/*
+ *  The set consists of SET_SIZE buckets, with each bucket containing 32 slots.
```
Why 32 slots?

qwe661234 replied:
I just set a sufficiently large number to prevent the bucket from becoming full.

jserv replied:
Add more comments and mark it as configurable. We shall always distinguish between magic numbers and configurations.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,435 @@
+ */
+static bool set_has(set_t *set, uint32_t key)
+{
+    uint32_t index = HASH(key);
```
Add `const` qualifier for `index`.

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,435 @@
+{
+    uint32_t index = HASH(key);
+    uint8_t count = 0;
+    while (set->table[index][count]) {
```
Rewrite as following:
```c
    for (uint8_t count = 0; set->table[index][count]; count++) {
```

jserv commented on src/compile.c:
```diff
@@ -0,0 +1,435 @@
+
+    if (!jit_code_string) {
+        jit_code_string = malloc(sizeof(code_string_t));
+        jit_code_string->curr = 0;
```
Move the statement `jit_code_string->curr = 0;` out of if-else block.

jserv commented on src/emulate.c:
```diff
@@ -1313,6 +211,7 @@ static inline bool insn_is_unconditional_branch(uint8_t opcode)
     return false;
 }
 
+#if !RV32_HAS(JIT)
```
Is it possible to unify hash function implementation? i.e., unify macro `HASH` and function `hash`.

qwe661234 replied:
It is possible, but it may impact the performance of interpreter.

jserv replied:
Add FIXME/TODO comments within this pull request. Then, submit an issue on GitHub to address code consolidation.

qwe661234 committed:
Introduce baseline JIT compiler

The baseline JIT compiler design involves tracing an EBB when its usage
frequency exceeds a predetermined threshold. We then utilize a code generator
to convert the instruction sequence into C code. The C code is subsequently
compiled using mir, and the resulting target machine code is stored in the
code cache for future utilization.

The primary objective of introducing the baseline JIT compiler is to enhance
the execution speed of RISC-V instructions with minimal modification. This
implementation requires three additional components: a code generator, mir
compiler, and code cache, while maintaining the original design of the interpreter.
Furthermore, this baseline JIT compiler serves as the foundational target
for future improvements.

In addition, we have developed a Python script that effectively traces code templates and automatically generates JIT code templates. This approach eliminates the need for manually writing duplicated codey.

* Core i7-8700
| Metric    |     master      |  baseline JIT   |Speedup |
|-----------+-----------------+-----------------+--------|
| CoreMark  |1352.843 (iter/s)|1932.496 (iter/s)| +42.8% |
| Dhrystone |1282 DMIPS       |2495.11 DMIPS    | +94.6% |

* eMAG
| Metric    |     master      |  baseline JIT   |Speedup |
|-----------+-----------------+-----------------+--------|
| CoreMark  | 438.073 (iter/s)| 644.426 (iter/s)| +47.1% |
| Dhrystone | 354.60 DMIPS    | 867.00 DMIPS    |  +144% |
